src/application/predictor.hpp

```diff
@@ -40,7 +41,7 @@ class Predictor {
   */
   Predictor(Boosting* boosting, int start_iteration, int num_iteration, bool is_raw_score,
             bool predict_leaf_index, bool predict_contrib, bool early_stop,
-            int early_stop_freq, double early_stop_margin) {
+            int early_stop_freq, double early_stop_margin, PredictionControlParameter& predict_params) {
```

Comment:
What's the benefit of introducing this new `PredictionControlParameter` class instead of just adding a new argument to the `Predictor`, just as we do for other customizations like `predict_leaf_index` and `predict_contrib`?

From my read of the code, it looks like unnecessary extra indirection and complexity, and I'm not sure what it gives us in exchange for that.

Reply:
If only considering random-assignment-prediction, a new parameter to Predictor of vector<int> is enough and more simple. What led me to introduce PredictionControlParameter is that I found there are many scattered code modification for adding a new parameter. I think introducing a structure of parameters could lower the code changes for the next variant of the prediction function. But I have to say that I have also wondered whether it is necessary. So If you guys finally decide to accept this feature and expect more clear and simple implementation, I could remove this design. :innocent: After all global replacement is not a difficult thing